src/nr_openai_observability/bedrock.py

```diff
@@ -0,0 +1,414 @@
+logger.addHandler(logging.StreamHandler(sys.stdout))
+logger.setLevel(logging.INFO)
+
+def perform_patch_bedrock():
+    from newrelic.common.package_version_utils import get_package_version_tuple
+
+    (major, minor, revision) = get_package_version_tuple('botocore')
+
+    if major != 1 and minor < 31 and revision < 57:
+        logger.warning(f'minimum version of botocore that supports Bedrock is 1.31.57')
+        return
+
+    (major, minor, revision) = get_package_version_tuple('boto3')
+
+    if major != 1 and minor < 28 and revision < 57:
+        logger.warning('minimum version of boto3 that supports Bedrock is 1.28.57')
+        return
+
+    try:
+        botocore.client.ClientCreator.create_client = _patched_call(
+            botocore.client.ClientCreator.create_client, patcher_aws_create_api
+        )
+    except AttributeError as error:
+        logger.debug(f'failed to instrument botocore.client.ClientCreator.create_client: {error}')
+
+
+def patcher_aws_create_api(original_fn, *args, **kwargs):
+    # We are in the AWS API to create a new object, we need to invoke that first!
+    try:
+        from botocore.model import ServiceModel
+        response = original_fn(*args, **kwargs)
+
+        cls = type(response)
+        name = cls.__qualname__
+        if cls.__module__ is not None and cls.__module__ != '__builtin__':
+            name = f'{cls.__module__}.{cls.__qualname__}'
+
+        if name == 'botocore.client.BedrockRuntime':
+            bedrock_method = 'invoke_model'
+            original_invoke_model = getattr(response, bedrock_method)
+
+            if original_invoke_model is not None:
+                setattr(response, bedrock_method, _patched_call(
+                    original_invoke_model, 
+                    patcher_bedrock_create_completion
+                ))
+            else:
+                logger.error(f"failed to find method '{bedrock_method}' on {response}")
+
+        return response
+
+    except BaseException as error:
+        logger.error(f'caught exception trying to invoke original function: {error}')
+        raise error
+
+
+def patcher_bedrock_create_completion(original_fn, *args, **kwargs):
+    timestamp = time.time()
+    result, time_delta = None, None
+
+    try:
+        with newrelic.agent.FunctionTrace(
+            name="AI/Bedrock/Chat/Completions/Create", group="", terminal=True
+        ):
+            monitor.record_library('botocore', 'Bedrock')
+            result = original_fn(*args, **kwargs)
+            time_delta = time.time() - timestamp
+    except Exception as error:
+        build_completion_summary_for_error(error, **kwargs)
+        raise result
+
+    # print the HTTP body
+    from botocore.response import StreamingBody
+    from io import BytesIO
+
+    contents = result['body'].read()
+    bio = BytesIO(contents)
+    result['body'] = StreamingBody(bio, len(contents))
+
+    handle_bedrock_create_completion(result, time_delta, **kwargs)
```

Comment:
Could we pass this `contents` here to the `handle_` method below and then only reset the body after the `handle_` call?

Reply:
Done. It still requires passing in both result and contents as the `handle_` function needs access to both.